library/core/src/mem/transmutability.rs

```diff
@@ -1,10 +1,88 @@
 use crate::marker::{ConstParamTy_, UnsizedConstParamTy};
 
-/// Are values of a type transmutable into values of another type?
+/// Marks that `Src` is transmutable into `Self`.
 ///
-/// This trait is implemented on-the-fly by the compiler for types `Src` and `Self` when the bits of
-/// any value of type `Self` are safely transmutable into a value of type `Dst`, in a given `Context`,
-/// notwithstanding whatever safety checks you have asked the compiler to [`Assume`] are satisfied.
+/// # Implementation
+///
+/// This trait cannot be implemented explicitly. It is implemented on-the-fly by
+/// the compiler for all types `Src` and `Self` such that, given a set of safety
+/// obligations on the programmer (see [`Assume`]), the compiler has proved that
+/// the bits of a value of type `Src` can be soundly reinterpreted as a `Self`.
+///
+/// # Safety
+///
+/// If `Dst: BikeshedIntrinsicFrom<Src, ASSUMPTIONS>`, the compiler guarantees
+/// that `Src` is soundly *union-transmutable* into a value of type `Dst`,
+/// provided that the programmer has guaranteed that the given
+/// [`ASSUMPTIONS`](Assume) are satisfied.
+///
+/// A union-transmute is any bit-reinterpretation conversion in the form of:
```

Comment:
I love how concise this is.

One question, though: are the semantics of this operation fully specified? In particular, if `size_of::<Src>() < size_of::<Dst>()`, is it well-defined that the trailing bytes in `Transmute<Src, Dst>` are uninitialized?

If the answer is yes, then no change is required here in response to this comment.

Reply:
It's how Rust works now, but it's not as well-documented as I'd like — we'll need to cite other official documentation before we stabilize this. That said, I'm confident the decision here is a 'safe' one. `MaybeUninit` relies on these semantics, though theoretically that could change. I think that's unlikely, since the present discussion is whether `union`s should *always* implicitly admit uninitialized bytes, even outside of the padding-extension case.

Should that discussion be rolled back so severely that Rust, say, documents those excess bytes to be initialized, we have a secondary bulwark in @scottmcm's plan to redefine the `transmute_unchecked` to permit padding extension in this same way (it's useful for SIMD).

This is all to say: I think the winds are blowing sufficiently hard in the right direction that we can tentatively adopt these semantics. We'll need to pin them down definitively before stabilization, but that's a long ways off.

Reply:
Sounds reasonable. Can we document this as a blocker somewhere to make sure it's resolved before stabilization?

Reply:
I've added a note to *Unresolved Questions* in the [tracking issue](https://github.com/rust-lang/rust/issues/99571).